archive/100-theorems-list/81_sum_of_prime_reciprocals_diverges.lean

```diff
@@ -0,0 +1,241 @@
+      split,
+      { intro p,
+        have h1 : p ∈ (m + 1).factors → ∃ a : ℕ, a < k ∧ a.succ = p,
+        { simp_intros hp only [(nat.mem_factors (nat.zero_lt_succ m))],
+          exact ⟨p.pred, (nat.pred_lt (nat.prime.ne_zero hp.left)).trans_le ((hmp p) hp),
+                nat.succ_pred_eq_of_pos (nat.prime.pos hp.left)⟩ },
+        simpa },
+      { have h2 : (m + 1).factors.prod - 1 = m,
+        { simpa only [nat.prod_factors (nat.zero_lt_succ m)] using nat.succ_sub_one m },
+        simp_rw [f], simpa } },
+    simpa },
+
+  calc card M₁ ≤ card (image f K)                    : card_le_of_subset h
+  ...          ≤ card K                              : card_image_le
+  ...          ≤ 2 ^ card (image nat.succ (range k)) : by simp only [K, card_powerset]
+  ...          ≤ 2 ^ card (range k)                  : pow_le_pow one_le_two card_image_le
+  ...          = 2 ^ k                               : by rw card_range k,
+end
+
+lemma card_M_le_2_pow_k_mul_sqrt_x {x k : ℕ} :
+  card {e ∈ range x | ∀ p : ℕ, (nat.prime p ∧ p ∣ (e + 1)) → p ≤ k} ≤ 2 ^ k * nat.sqrt x :=
+begin
+  let M := {e ∈ range x | ∀ p : ℕ, (nat.prime p ∧ p ∣ (e + 1)) → p ≤ k},
+  let M₁ := {e ∈ range x | squarefree (e + 1) ∧ ∀ p : ℕ, (nat.prime p ∧ p ∣ (e + 1)) → p ≤ k},
+  let M₂ := {e ∈ range (nat.sqrt x) | ∀ p : ℕ, (nat.prime p ∧ p ∣ (e + 1)) → p ≤ k},
+  let K := finset.product M₁ M₂,
+  let f : ℕ × ℕ → ℕ := λ mn, (mn.2 + 1) ^ 2 * (mn.1 + 1) - 1,
+
+  have h1 : M ⊆ image f K,
+  { intros m hm,
+    simp only [M, M₁, M₂, mem_image, exists_prop, prod.exists, mem_product, sep_def, mem_filter,
+               mem_range] at hm ⊢,
+    have hm' := m.zero_lt_succ,
+    obtain ⟨a, b, hab₁, hab₂⟩ := nat.sq_mul_squarefree_of_pos' hm',
+    obtain ⟨ham, hbm⟩ := ⟨dvd.intro_left _ hab₁, dvd.intro _ hab₁⟩,
+    refine ⟨a, b, ⟨⟨_, hab₂, λ p hp, _⟩, ⟨_, λ p hp, _⟩⟩, by simp_rw [f, hab₁, m.succ_sub_one]⟩,
+    { exact (nat.succ_le_succ_iff.mp (nat.le_of_dvd hm' ham)).trans_lt hm.1 },
+    { exact hm.2 p ⟨hp.1, dvd.trans hp.2 ham⟩ },
+    { calc b < b + 1        : lt_add_one b
+      ...    ≤ (m + 1).sqrt : by simpa only [nat.le_sqrt, pow_two] using nat.le_of_dvd hm' hbm
+      ...    ≤ x.sqrt       : nat.sqrt_le_sqrt (nat.succ_le_iff.mpr hm.1) },
+    { exact hm.2 p ⟨hp.1, dvd.trans hp.2 (nat.dvd_of_pow_dvd one_le_two hbm)⟩ } },
+
+  have h2 : card M₂ ≤ nat.sqrt x,
+  { rw ← card_range (nat.sqrt x), apply card_le_of_subset, simp [M₂] },
+
+  calc card M ≤ card (image f K)   : card_le_of_subset h1
+  ...         ≤ card K             : card_image_le
+  ...         = card M₁ * card M₂  : card_product M₁ M₂
+  ...         ≤ 2 ^ k * x.sqrt     : mul_le_mul' card_M1_le_2_pow_k h2,
+end
+
+theorem real.tendsto_sum_one_div_prime_at_top :
+  tendsto (λ n, ∑ p in {p ∈ range n | nat.prime p}, (1 / (p : ℝ))) at_top at_top :=
+begin
+  by_contradiction,
+  obtain ⟨k, h1⟩ := not_tendsto_imp_sum_lt_half h,
```

Comment:
Since you call `h` by name it is probably preferable if you introduce it explicitly.
```suggestion
  by_contradiction h,
  obtain ⟨k, h1⟩ := not_tendsto_imp_sum_lt_half h,
```
Also, just as an fyi, you can use `by_contra` as shorthand for `by_contradiction`.